### PR TITLE
Add CHERI ABI TLS

### PR DESCRIPTION
**Pull request by jrtc27**

### CODE REVIEW

**Comment by brooksdavis** [comment]
Looks good to me.

**Comment by jrtc27** [comment]
I had to rebase this and for some reason I'm now getting an assertion failure in LLVM when compiling `jemalloc_extents.c` for CheriABI. Currently debugging/reducing so please hold off on merging until that's been resolved (it looks like my LLVM CHERI intrinsics folding changes may be at fault).

File: libexec/rtld-cheri-elf/rtld.c [hunk 1]
```diff
@@ -4710,48 +4710,109 @@ tls_get_addr_common(intptr_t **dtvp, int index, size_t offset)
...
+    tls_block_size = tcbsize + post_size;
+    pre_size = roundup2(tls_block_size, tls_init_align) - tls_block_size;
+
+    return ((char *)tcb - pre_size - extra_size);
+}
+
 /*
  * Allocate Static TLS using the Variant I method.
+ *
+ * To handle all above requirements, we setup the following layout for
```
Comment (bsdjhb) [inline libexec/rtld-cheri-elf/rtld.c]: The requirements aren't listed here (they are in the libc source), so I would perhaps just add a pointer to the comments in lib/libc/gen/tls.c.

File: libexec/rtld-elf/rtld.c [hunk 2]
```diff
@@ -4709,48 +4709,89 @@ tls_get_addr_common(Elf_Addr **dtvp, int index, size_t offset)
...
+#endif
+    tls_block_size = tcbsize + post_size;
+    pre_size = roundup2(tls_block_size, tls_init_align) - tls_block_size;
+
+    return ((char *)tcb - pre_size - extra_size);
+}
+
 /*
  * Allocate Static TLS using the Variant I method.
  */
```
Comment (bsdjhb) [inline libexec/rtld-elf/rtld.c]: You did not add the comment here in the real rtld, and I would add the same "see also" reference to lib/libc/gen/tls.c.

**Comment by jrtc27** [comment]
RTLD changes forwarded upstream at https://reviews.freebsd.org/D16510.

**Pull request merged**